File: examples/scala/src/main/scala/com/xebia/functional/xef/scala/auto/Book.scala
package com.xebia.functional.xef.scala.conversation

import com.xebia.functional.xef.scala.conversation.Description
import com.xebia.functional.xef.scala.conversation.*
import io.circe.Decoder
import com.xebia.functional.xef.prompt.Prompt

@Description("A book")
case class Book(
    @Description("the name of the book") name: String,
    @Description("the author of the book") author: String,
    @Description("A 50 word paragraph with a summary of this book") summary: String
) derives SerialDescriptor,
      Decoder

def summarizeBook(title: String, author: String)(using conversation: ScalaConversation): Book =
  prompt(Prompt(s"$title by $author summary."))

@main def runBook: Unit =
  conversation {
    val toKillAMockingBird = summarizeBook("To Kill a Mockingbird", "Harper Lee")
    println(s"${toKillAMockingBird.name} by ${toKillAMockingBird.author} summary:\n ${toKillAMockingBird.summary}")
  }

File: examples/kotlin/src/main/kotlin/com/xebia/functional/xef/auto/fields/Book.kt
package com.xebia.functional.xef.auto.fields

import com.xebia.functional.xef.auto.Description
import com.xebia.functional.xef.auto.conversation
import com.xebia.functional.xef.auto.llm.openai.OpenAI
import com.xebia.functional.xef.auto.llm.openai.prompt
import kotlinx.serialization.Serializable

@Serializable
data class Book(
  @Description("The title of the book.") val title: String,
  @Description("The author of the book.") val author: String,
  @Description("A 50 word summary of the book.") val summary: String
)

suspend fun main() {
  OpenAI.conversation {
    val toKillAMockingbird: Book = prompt("To Kill a Mockingbird by Harper Lee")
    println(toKillAMockingbird)
  }
}
